data/modules/SecondHand/SecondHand.lua

```diff
@@ -14,116 +14,166 @@ local PlayerState = require 'PlayerState'
-local l2 = Lang.GetResource("ui-core")
 
 -- average number of adverts per BBS and billion population
 local N_equil = 0.1 -- [ads/(BBS*unit_population)]
 
 -- inverse half life of an advert in (approximate) hours
-local inv_tau = 1.0 / (4*24)
+local inv_tau = 1.0 / (4 * 24)
 
+-- Maximum indeces for randomised messages; see data/lang/module-secondhand/en.json
 -- Note: including the 0
 local max_flavour_index = 5
 local max_surprise_index = 2
+local max_money_index = 3
 
 local flavours = {}
-for i = 0,max_flavour_index do
-    table.insert(flavours, {
-        adtitle = l["FLAVOUR_" .. i .. "_TITLE"],
-        adtext = l["FLAVOUR_" .. i .. "_TEXT"],
-        adbody  = l["FLAVOUR_" .. i .. "_BODY"],
-    })
+for i = 0, max_flavour_index do
+	table.insert(flavours, {
+		adtitle = l["FLAVOUR_" .. i .. "_TITLE"],
+		adtext  = l["FLAVOUR_" .. i .. "_TEXT"],
+		adbody  = l["FLAVOUR_" .. i .. "_BODY"],
+	})
 end
 
 local ads = {}
 
 -- find a value in an array and return key
-local onDelete = function (ref)
-    ads[ref] = nil
+local onDelete = function(ref)
+	ads[ref] = nil
 end
 
--- check it fits on the ship, both the slot for that equipment and the
--- weight.
+-- Check if a piece of equipment can fit into the ship. Both the available space
+-- and, for equipemnt which requires a slot, a free and compatible slot are
+-- checked.
+--
+-- Returns a triple of:
+--
+--   ok      - whether the equipment can be installed
+--
+--   slot?   - if ok and the equipment requires a slot, the free slot into which
+--             it can be installed
+--
+--   message - a message containing the reason why the equipment can't be installed
+--
 ---@param e EquipType
-local canFit = function (e)
-
+---@return boolean,Hullconfig.Slot?,string
+local canFit = function(e)
 	local equipSet = Game.player:GetComponent("EquipSet")
+	local hasEnoughFreeVolume = equipSet:GetFreeVolume() >= e.volume
+
+	if not e.slot then
+		return hasEnoughFreeVolume, nil, l.EQUIPMENT_NO_SPACE
+	end
 
-	if e.slot then
-		local slot = equipSet:GetFreeSlotForEquip(e)
-		return slot ~= nil, slot, l2.SHIP_IS_FULLY_EQUIPPED
-	else
-		return equipSet:CanInstallLoose(e), nil, l2.SHIP_IS_FULLY_EQUIPPED
+	if not equipSet:HasCompatibleSlotForEquipment(e) then
+		return false, nil, l.EQUIPMENT_NO_COMPATIBLE_SLOT
 	end
+	if not hasEnoughFreeVolume then
+		return false, nil, l.EQUIPMENT_SLOT_NO_SPACE
+	end
+	local slot = equipSet:GetFreeSlotForEquip(e)
+	return slot ~= nil, slot, l.EQUIPMENT_NO_FREE_SLOT
 end
 
+-- Check if the player can afford the equipment item for sale
+local canAfford = function(ad)
+	return PlayerState.GetMoney() >= ad.price
+end
 
-local onChat = function (form, ref, option)
-    local ad = ads[ref]
+-- Return true if the ad should be enabled in the BBS
+local isEnabled = function(ref)
+	if ads[ref] == nil then return false end
+	if not canAfford(ads[ref]) then return false end
+	if not canFit(ads[ref].equipment) then return false end
+	return true
```

Comment:
Nitpick: rather than multiple early-returns, prefer to use short-circuit evaluation for trivial conditions like this:

```suggestion
	return ads[ref] and canAfford(ads[ref]) and canFit(ads[ref].equipment)
```

Reply:
In general I also prefer a single return, but in this case I made it multi-return because:

- this is effectively "template" code and having it like this makes it more generic (each condition is by itself; some can be reused by other modules, others deleted/replaced)
- it makes it easy to add/remove conditions
- IMHO it makes it easier to see what each individual condition is rather than having a huge long line (easier to read multiple independant conditions top-to-bottom than left-to-right)

.. but changed to fit your preference. (squashed)

---

PS: it would be easier to have all this feedback in the original review so it can all be addressed once instead of having to iterate over this for days..

Reply:
I agree wholeheartedly regarding continued iteration - I apologize for the delay in getting this feedback in, but the initial batch was a 4am review and there were some things I didn't catch.

Regarding the template-like strategy - I also prefer to do it that way once the number of conditions scales past 3-4; at the current number of parameters it's not a significant hassle to refactor into a multi-return form-factor and so I prefer to use the shorter variant where possible (less keyword noise means it's a lot easier to read and understand).

Reply:
> I agree wholeheartedly regarding continued iteration - I apologize

No worries; sorry for mentioning it..